data_structures/flag_filter.wdl

```diff
@@ -0,0 +1,157 @@
+## and `exclude_if_all`.
+## These four fields correspond to the samtools flags
+## `-f`, `-F`, `--rf`, and `-G` respectively.
+## The values of these fields are strings that represent a 12bit bitwise flag.
+## These strings must evaluate to an integer less than 4096 (2^12).
+## They can be in octal, decimal, or hexadecimal format.
+## Please see the `meta.help` of `validate_string_is_oct_dec_or_hex`
+## for more information on the valid formats.
+##
+## The `validate_FlagFilter` workflow can be used to validate a `FlagFilter` struct.
+## **WARNING** The `validate_FlagFilter` workflow will only check that all the fields
+## can be parsed as integers less than 4096. It will not check if the flags are
+## sensible input to `samtools fastq`.
+## `samtools fastq` also employs very little error checking on the flags.
+## So it is possible to pass in flags that produce nonsensible output.
+## For example, it is possible to pass in flags that produce no output.
+## **Please exhibit caution while modifying any default values of a `FlagFilter`.**
+##
+## We suggest using the Broad Institute's SAM flag explainer to construct the flags.
+## Find it [here](https://broadinstitute.github.io/picard/explain-flags.html).
+##
+## ## Example input JSON
+##
+## ```json
+## {
+##    "flags": {
+##        "include_if_all": "0x3",
+##        "exclude_if_any": "0xF04",
+##        "include_if_any": "0x0",
+##        "exclude_if_all": "0x0"
+##    }
+## }
+## ```
+##
+## ### Explanation
+##
+## The above example JSON represents a `FlagFilter` struct
+## being passed to parameter named `flags`.
+## The `include_if_all` field is set to `0x3` which is `3` in decimal.
+## The `exclude_if_any` field is set to `0xF04` which is `3844` in decimal.
+## The `include_if_any` field is set to `0x0` which is `0` in decimal.
+## The `exclude_if_all` field is set to `0x0` which is `0` in decimal.
+##
+## `3` in decimal can be represented as `000000000011` in 12bit binary.
+## This number means that to be included a read must have the 1st and 2nd bits set.
+## Those bits correspond to the `read paired` and `read mapped in proper pair` flags.
+##
+## `3844` in decimal can be represented as `111100000100` in 12bit binary.
+## This number means that to be excluded a read must have **any** of the
+## 3rd, 9th, 10th, 11th, or 12th bits set.
+## We won't go through what all those bits mean here, but you can find
+## the meanings of the bits in the
+## [SAM flag explainer](https://broadinstitute.github.io/picard/explain-flags.html).
+## In short, those are all flags corresponding to the quality of the read
+## and them being `true` may indicate that the read is of low quality and
+## should be excluded.
+
+version 1.1
+
+struct FlagFilter {
+    String include_if_all  # samtools -f
+    String exclude_if_any  # samtools -F
+    String include_if_any  # samtools --rf
+    String exclude_if_all  # samtools -G
+}
+
+task validate_string_is_oct_dec_or_hex {
```

Comment:
I'm not sure that we'd ever use this anywhere else, but this doesn't necessarily have to be specific to this data structure.

Reply:
I started typing a disagreement with you, but I came around. This can be abstracted some so it can be re-used (hypothetically) and I think that's a good idea, but where would it belong? I feel like it kinda has to go into a file of its own, which seems a little silly. And I also wouldn't call it a data structure, which means we'd need... `validation/` in the root???
I'm not sure what to do with it 🤷